### PR TITLE
CI: Cache build improves build time from 5mins to 1m30s-2m10s

### PR DESCRIPTION
**Pull request by justin-lovell**
I got annoyed by the build times on the CI. 

So here is a fix.

image

### CODE REVIEW

**Comment by rurounijones** [comment]
Would it be more future-proof to use this rust specific cache action? https://github.com/marketplace/actions/rust-cache

**Comment by justin-lovell** [comment]
I did come across that package - and it should work too. The only advantage it has over the builtin cache is the following:

> Additionally, the action automatically works around cargo#8603 / actions/cache#403 which would otherwise corrupt the cache _on macOS builds._

Since we aren't targeting that platform, it's kind of a moot point.

The primary motivation for using the built-in cache action is that no needed to review the contents of the action in depth. I personally try to avoid absorbing dependencies unless there is a distinct advantage. Otherwise it becomes the `npm`-wild-west :) 

Having said that, I am not adverse to that option either for this utility function.

Let me know what your preference would be -- I think the cache will help a lot in the feedback loop.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -19,6 +19,16 @@ jobs:
       - name: Checkout code
         uses: actions/checkout@v2
 
+      - name: Cache
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.cargo/registry
+            ~/.cargo/git
+            target
```
Comment (rkusa) [inline .github/workflows/ci.yml:28]: The [cargo example](https://github.com/actions/cache/blob/main/examples.md#rust---cargo) has slightly different entries, I am wondering whether we should go with them instead? It is probably the same, as the example only narrows it down to subdirectories:

```suggestion
            ~/.cargo/bin/
            ~/.cargo/registry/index/
            ~/.cargo/registry/cache/
            ~/.cargo/git/db/
            target/
```

**Review by rkusa — COMMENTED** [review]
I am in favour of directly using the official `actions/cache` rather than an action (dependency) from a random user from the marketplace. I'd consider copy-and-paste easy enough here to not pull in another dependency.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -19,6 +19,16 @@ jobs:
...
+          key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+            ${{ runner.os }}-cargo
```
[the remark this answers is not in the document]
Comment (justin-lovell) [reply .github/workflows/ci.yml:30]: The indent on line 30 triggers a "continue as if on same line". They call it "Multi-line Flow Scalars"

File: .github/workflows/ci.yml (hunk 1 above)
Comment (justin-lovell) [reply .github/workflows/ci.yml:28]: I didn't find that example. Great find. 

I will align to that example if @rurounijones agrees to keep it simplistic on the dependency chain.

File: .github/workflows/ci.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (rkusa) [reply .github/workflows/ci.yml:30]: Gotcha, thanks for the clarification. If both is basically one line, then I think `${{ runner.os }}-cargo` is duplicated and one can be removed?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (rurounijones) [reply .github/workflows/ci.yml:28]: Yeah, lets just copy/paste and not add the dependency

File: .github/workflows/ci.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (justin-lovell) [reply .github/workflows/ci.yml:30]: Yeah, give we are not going for multi-target ... it does "simplify" the line at the expense of being "comprehensive. "

File: .github/workflows/ci.yml (hunk 1 above)
Comment (justin-lovell) [reply .github/workflows/ci.yml]: Sorted and pushed

**Pull request merged**